### ISSUE TITLE
Extremely slow speeds with tethered laptop

### ISSUE THREAD

**Issue by Minion-of-Cthulhu** [pyamsoft/tetherfi#315]
First of all, I _love_ the app. It has worked flawlessly, up until yesterday (June 30). I'm not sure if it's related, but I updated the app to the latest release (43, I believe) and now I get horribly slow speeds on my laptop that I tether to my phone but the speeds on the phone itself are fine.

According to several speed test sites I've tried, my phone averages about 15 Mbps. When I run a speed test from my laptop I get around 300 kbps with a latency around 2.5 ms for download and jitter of around 416 ms for download. I have a network traffic monitor on my phone which more or less consistently shows roughly 200 - 300 kbps when I'm using the laptop but will show normal speeds if I load a web page on my phone. The speed will sometimes bump up to 1.5 Mbps or so when I load a page with the laptop, and occasionally as high as 8.5 Mbps, but it's not stable and generally bounces wildly between 1 kbps and 3 Mbps, though the actual speed on the laptop is always extremely slow to the point where it can take 10 - 15 seconds for Google search results to load and up to a minute for something image heavy like YouTube to load. Interestingly, YouTube videos will stream fine on the laptop but, again, it takes forever for the page to actually load.

I also occasionally get pages that fail to load on the laptop and give me the following error, which I got when trying to load a Google search:

This site can’t provide a secure connection www.google.com sent an invalid response.
Try running Connectivity Diagnostics.
ERR_SSL_PROTOCOL_ERROR

That's obviously not an accurate error message, and refreshing the page works fine though it's absurdly slow. I've also seen other similar errors about sites not being available, etc. which are never real errors since a refresh allows the page to load.

I have the following Tetherfi settings turned on:

* Show Hotspot Notification
* 5 GHz
* Keep Wifi Awake
* Keep CPU Awake
* Stop Hotspot with No Clients
* Stubborn Proxy

Those are the same settings I was previously using. I've never been able to get the Always Alive to work, nor the Power Balance, Enable Idle Timer, or Avoid VPN Blocker Dialog.  I assume my phone, a Samsung A54, doesn't support them but, as I said, I wasn't using them before so I doubt they're the problem.

I also haven't changed anything on my laptop, and haven't had anything updated recently and certainly not yesterday when things stopped working well. The connection settings in Tetherfi are the defaults, and the connection settings on the laptop haven't changed. It connects to the phone fine, though occasionally it's a little slow to connect. It's really just that the laptop speeds are slow.

I've tried some standard troubleshooting:

* I shut down both the phone and laptop, waited several minutes, and booted them back up and tethered the laptop to the phone. I've also restarted both, rather than shutting them down. I've done both shut downs and restarts multiple times.
* I've turned airplane mode on, waited, then turned it back off. Sometimes this seems to give a slight speed increase on the laptop, but I can't tell if it really is faster or if it just feels that way.
* I've reset the phone's network settings.
* I uninstalled Tetherfi, restarted the phone, and reinstalled Tetherfi from the Play Store.
* I've also run the speed test on my laptop with Tetherfi open on my phone screen, as I though maybe it was some sort of issue with the app sleeping or some such, but that made no difference. 

None of those have fixed the problem or made any improvements.

Any ideas? If you need a copy of the developer logs, let me know.

**Comment by pyamsoft** [comment]
Hey, sorry to hear things are slow for you.

The only thing I could think would be that your phone providing the hotspot is slowing down because its doing "too much work". It is for cases just like these that the "Enable Idle Timeout" setting was added, as it will clean up bad connections.

Depending on your phone, you may find improvement by enabling Always Alive (which is the same as changing the Batter Optimization settings for the TetherFi app to "Unoptimized"). Also, it could be that since the new update was just recently installed, your device hasn't had enough time to generate baseline profiles for it, which can influence behavior somewhat.

Unfortunately, it sounds like since you are hitting these SSL protocol errors, this issue is outside of the scope of TetherFi itself - unless something else has gone horribly wrong with the hotspot proxy code. TetherFi cannot control anything about the speed of your connection, so anything network speed related is sadly out of my control :(

**Comment by pyamsoft** [comment]
Also, please try the 2.4 band instead of the 5 band, and turn off the "Stop Hotspot" tweak.
What kind of device do you have?

**Comment by Minion-of-Cthulhu** [comment]
I've got a Samsung A54. Like I said, everything worked fine until I updated the app on Sunday.

**Comment by pyamsoft** [comment]
OK, so its not like your device is too old or too slow.

Unfortunately I have basically no idea. Yes the proxy will always be a little slower than your device by itself, but the proxy code has largely not changed for many releases now.

I can't point to anything specific as a reason for why your device would suddenly be slower now even though it was not before, except for purely coincidental situations, sorry :(

**Comment by Minion-of-Cthulhu** [comment]
I appreciate the quick response!

I tried the 2.4 GHz band and saw a minor improvement to about 500 kbps.

I turned on the Enable Idle Timeout and turned off the Stop Hotspot tweak and ran the test again. I got an average speed of 4.37 Mbps.

Not sure if that's useful or not.

Any other thoughts or suggestions?

**Comment by Minion-of-Cthulhu** [comment]
The previous post was done on the 5 GHz band. Once I switch back to 2.4 GHz and tested the speed it was back to normal, about 14 Mbps.

I'm really wondering why all this happened since I didn't change anything and you said the proxy code hasn't really changed.

**Comment by pyamsoft** [comment]
My one thought is related to the new idle connection timeout feature 


To keep a complex topic simple, some websites ask for a bunch of stuff like images, but never tell the server that they are done loading, so there's just a bunch of wasted internet as the website and the server keep waiting forever for the other one to talk. This behavior is basically random, and so maybe a string of bad luck got you stuck in this "slow mode".

Enabling the new tweak let's my code clean up these ugly connections, which frees up room for the connections that are actually moving data.

Basically, its all random, and today we got lucky! Sometimes that's life :)

**Comment by Minion-of-Cthulhu** [comment]
Well, whatever the reason, thanks for the help! I really appreciate it, and maybe this will help anyone else who might come by here with the same problem.

**Comment by Lq9000** [comment]
Lq9000905@gmail.com

**Comment by Lq9000** [comment]
https://www.google.com.br
